tmt/steps/prepare/artifact/providers/repository.py

```diff
@@ -2,125 +2,166 @@
-    Sets up a repository on the guest and lists the RPMs it provides.
+    Provider for making RPM artifacts from a repository discoverable without downloading them.
 
-    The artifact_id is a URL to a .repo file. This provider's main
-    purpose is to download this file and place it in '/etc/yum.repos.d/'.
+    The provider identifier should start with 'repository-url:' followed by a URL to a .repo file,
+    e.g., "repository-url:https://download.docker.com/linux/centos/docker-ce.repo".
 
-    It then lists all RPMs made available by the new repository but does
-    not download any of them.
+    Artifacts are all available RPM packages listed in the repository
     """
 
-    def __init__(self, raw_provider_id: str, logger: tmt.log.Logger):
-        super().__init__(raw_provider_id, logger)
-        self.repo_file = RepositoryFile(url=self.id)
-        # Cache for the list of RPMs discovered in the repository
-        self._rpm_list: list[RpmArtifactInfo] = []
+    repository: Repository
+    _artifact_list: Optional[Sequence[RpmArtifactInfo]]
 
     @classmethod
     def _extract_provider_id(cls, raw_provider_id: str) -> ArtifactProviderId:
-        return raw_provider_id
+        prefix = 'repository-url:'
+        if not raw_provider_id.startswith(prefix):
+            raise ValueError(f"Invalid repository provider format: '{raw_provider_id}'.")
+        value = raw_provider_id[len(prefix) :]
+        if not value:
+            raise ValueError("Missing repository URL.")
+        return value
 
-    def _fetch_rpms(self, guest: Guest, repo_filepath: Path) -> None:
-        """
-        Query the guest to find all packages available in the new repository.
-        """
-        # TODO: This method needs to be implemented to populate self._rpm_list with RPMs
-        # from the repository. Currently using an empty list as a temporary solution.
-        self._rpm_list = []
+    def __init__(self, raw_provider_id: str, logger: tmt.log.Logger):
+        super().__init__(raw_provider_id, logger)
+        # Initialize to None to distinguish between "not run" and "run but empty"
+        self._artifact_list = None
 
-    @property
+    @cached_property
     def artifacts(self) -> Sequence[RpmArtifactInfo]:
-        raise NotImplementedError
+        # Check for None to see if fetch_contents() has been called
+        if self._artifact_list is None:
+            raise tmt.utils.GeneralError("Call fetch_contents first to discover artifacts.")
+        # Return the list (which is valid even if it's empty)
+        return self._artifact_list
 
     def _download_artifact(
-        self, artifact: RpmArtifactInfo, guest: Guest, destination: Path
+        self, artifact: RpmArtifactInfo, guest: Guest, destination: tmt.utils.Path
     ) -> None:
-        """This provider only sets up the repo, it does not download RPMs."""
+        """
+        This provider does not support downloading individual artifacts.
+        """
+        raise DownloadError(
```

Comment:
```suggestion
        raise AssertionError(
```
I believe you mean here that `_download_artifact` should not even be called in normal  flow?